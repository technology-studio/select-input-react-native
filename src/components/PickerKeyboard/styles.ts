import { StyleSheet } from 'react-native'

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'flex-end',
  },
  pickerview: {
    backgroundColor: '#FFFFFF',
  },
})

export default styles
