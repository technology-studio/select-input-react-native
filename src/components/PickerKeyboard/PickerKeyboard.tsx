import React, {
  Component, type ComponentRef,
} from 'react'
import {
  Dimensions, type StyleProp, type EmitterSubscription,
  type TextStyle,
  type ViewStyle,
} from 'react-native'
import { Picker } from '@react-native-picker/picker'

import { CustomKeyboard } from '../CustomKeyboard'
import { type Option } from '../AbstractSelectInput/AbstractSelectInput'

import { styles } from './styles'

type Props = {
  buttonsTextStyle?: StyleProp<TextStyle>,
  buttonsViewStyle?: StyleProp<ViewStyle>,
  cancelKeyText: string,
  onCancel: () => void,
  onSubmit: (value: string) => void,
  onValueChange: (value: string) => void,
  options: Option[],
  pickerItemStyle?: StyleProp<ViewStyle>,
  pickerViewStyle?: StyleProp<ViewStyle>,
  submitKeyText: string,
  value: string,
}

type State = {
  value: string,
  visible: boolean,
  width: number,
}

export class PickerKeyboard extends Component<Props, State> {
  picker: ComponentRef<typeof Picker<string>> | null
  dimensionsListener: EmitterSubscription | undefined
  constructor (props: Props) {
    super(props)

    this.picker = null
    this.state = {
      value: props.value,
      visible: false,
      width: Dimensions.get('window').width,
    }
  }

  componentDidMount (): void {
    this.dimensionsListener = Dimensions.addEventListener('change', this.updateDimensions)
  }

  componentWillUnmount (): void {
    this.dimensionsListener?.remove()
  }

  updateDimensions = (): void => {
    this.setState({
      width: Dimensions.get('window').width,
    })
  }

  setPickerRef = (component: ComponentRef<typeof Picker<string>> | null): void => {
    this.picker = component
  }

  componentDidUpdate (prevProps: Props): void {
    const { value } = this.props
    const prevValue = prevProps.value

    if (value !== prevValue) {
      this.setState({
        value,
      })
    }
  }

  focus (): void {
    this.setVisible(true)
  }

  onCancelPress = (): void => {
    const { onCancel } = this.props

    this.setVisible(false)
    onCancel?.()
  }

  onSubmitPress = (): void => {
    const { onSubmit } = this.props
    const { value } = this.state

    this.setVisible(false)
    onSubmit?.(value)
  }

  onValueChange = (value: string): void => {
    const { onValueChange } = this.props
    onValueChange?.(value)

    this.setState({
      value,
    })
  }

  setVisible = (visible: boolean): void => {
    this.setState({
      visible,
    })
  }

  render (): JSX.Element {
    const { value, visible, width } = this.state
    const {
      buttonsTextStyle,
      buttonsViewStyle,
      pickerItemStyle,
      pickerViewStyle,
      cancelKeyText,
      submitKeyText,
      options,
    } = this.props

    return (
      <CustomKeyboard
        buttonsTextStyle={buttonsTextStyle}
        buttonsViewStyle={buttonsViewStyle}
        cancelKeyText={cancelKeyText}
        onCancelPress={this.onCancelPress}
        onSubmitPress={this.onSubmitPress}
        submitKeyText={submitKeyText}
        visible={visible}
      >
        <Picker<string>
          ref={this.setPickerRef}
          onValueChange={this.onValueChange}
          selectedValue={value}
          style={[styles.pickerview, pickerViewStyle, { width }]}
          itemStyle={pickerItemStyle}
        >
          {options.map(option => (
            <Picker.Item
              key={option.value}
              value={option.value}
              label={option.label}
            />
          ))}
        </Picker>
      </CustomKeyboard>
    )
  }
}
