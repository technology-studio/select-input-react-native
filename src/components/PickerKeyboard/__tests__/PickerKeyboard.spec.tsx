import React from 'react'
import renderer, { type ReactTestInstance } from 'react-test-renderer'

import PickerKeyboard from '../PickerKeyboard'

const props = {
  buttonsTextStyle: {
    color: '#BBBBBB',
    fontSize: 13,
  },
  buttonViewStyle: {
    backgroundColor: '#FFFFFF',
    borderColor: '#AAAAAA',
    bottomBorderWidth: 0,
  },
  pickerViewStyle: {
    backgroundColor: '#CCCCCC',
  },
  cancelKeyText: 'cancelKeyText',
  onCancel: jest.fn(),
  onSubmit: jest.fn(),
  onValueChange: jest.fn(),
  options: [{ value: '0', label: '0' }],
  style: {},
  submitKeyText: 'submitKeyText',
  value: '0',
}

describe('PickerKeyboard', () => {
  describe('rendering', () => {
    // TODO: resolve issue with missing views in snapshot testing
    // test('renders correctly with given props', () => {
    //   const pickerKeyboard = renderer.create(<PickerKeyboard {...props} />)
    //   expect(pickerKeyboard).toMatchSnapshot()
    // })
  })

  describe('functions', () => {
    let pickerKeyboard: ReactTestInstance | null = null

    beforeEach(() => {
      pickerKeyboard = renderer
        .create(<PickerKeyboard {...props} />)
        .getInstance()
    })

    afterEach(() => {
      pickerKeyboard = null
    })

    test('sets picker ref correctly', () => {
      // @ts-expect-error -- TODO: fix type
      pickerKeyboard?.setPickerRef(1)
      // @ts-expect-error -- TODO: fix type
      expect(pickerKeyboard?.picker).toBe(1)
    })

    test('updates value correctly', () => {
      // @ts-expect-error state probably doesn't exist - check
      expect(pickerKeyboard?.state.value).toBe('0')
      // @ts-expect-error -- TODO: fix type
      pickerKeyboard?.componentDidUpdate({ value: '1' })
      // @ts-expect-error state probably doesn't exist - check
      expect(pickerKeyboard?.state.value).toBe('0')
    })

    test('updates visible state when focussing correctly', () => {
      // @ts-expect-error state probably doesn't exist - check
      expect(pickerKeyboard?.state.visible).toBe(false)
      // @ts-expect-error -- TODO: fix type
      pickerKeyboard?.focus()
      // @ts-expect-error state probably doesn't exist - check
      expect(pickerKeyboard?.state.visible).toBe(true)
    })

    test('updates value correctly when value changes', () => {
      // @ts-expect-error state probably doesn't exist - check
      expect(pickerKeyboard?.state.value).toBe('0')

      // @ts-expect-error -- TODO: fix type
      pickerKeyboard?.onValueChange('1')
      // @ts-expect-error state probably doesn't exist - check
      expect(pickerKeyboard?.state.value).toBe('1')
    })

    test('updates visibily correctly', () => {
      // @ts-expect-error state probably doesn't exist - check
      expect(pickerKeyboard?.state.visible).toBe(false)

      // @ts-expect-error -- TODO: fix type
      pickerKeyboard?.setVisible(true)
      // @ts-expect-error state probably doesn't exist - check
      expect(pickerKeyboard?.state.visible).toBe(true)

      // @ts-expect-error -- TODO: fix type
      pickerKeyboard?.setVisible(false)
      // @ts-expect-error state probably doesn't exist - check
      expect(pickerKeyboard?.state.visible).toBe(false)
    })
  })

  describe('callbacks', () => {
    const onCancel = jest.fn()
    const onSubmit = jest.fn()

    const pickerKeyboard = renderer
      .create(
        <PickerKeyboard {...props} onCancel={onCancel} onSubmit={onSubmit} />,
      )
      .getInstance()

    test('call onCancel prop correctly', () => {
      // @ts-expect-error -- TODO: fix type
      pickerKeyboard?.onCancelPress()

      expect(onCancel).toHaveBeenCalled()
      // @ts-expect-error state probably doesn't exist - check
      expect(pickerKeyboard?.state.visible).toBe(false)
    })

    test('call onSubmit prop correctly', () => {
      // @ts-expect-error -- TODO: fix type
      pickerKeyboard?.onSubmitPress()

      expect(onSubmit).toHaveBeenCalledWith('0')
      // @ts-expect-error state probably doesn't exist - check
      expect(pickerKeyboard?.state.visible).toBe(false)
    })
  })
})
