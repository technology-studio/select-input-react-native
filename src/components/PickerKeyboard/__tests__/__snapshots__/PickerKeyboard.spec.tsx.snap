// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PickerKeyboard rendering renders correctly with given props 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onOrientationChange={[Function]}
  supportedOrientations={
    Array [
      "portrait",
      "landscape",
      "landscape-left",
      "landscape-right",
    ]
  }
  transparent={true}
  visible={false}
>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "flex-end",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#F5FCFF",
            "justifyContent": "center",
            "padding": 0,
          },
          Object {
            "width": 750,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#CCCFD6",
              "borderBottomWidth": 0,
              "borderColor": "#CCCFD6",
              "borderTopColor": "lightgrey",
              "borderTopWidth": 0.5,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "padding": 0,
            },
            undefined,
            Object {
              "width": 750,
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#006BFF",
                  "padding": 8,
                },
                Object {
                  "color": "#BBBBBB",
                  "fontSize": 13,
                },
                Object {
                  "textAlign": "left",
                },
              ]
            }
          >
            cancelKeyText
          </Text>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#006BFF",
                  "padding": 8,
                },
                Object {
                  "color": "#BBBBBB",
                  "fontSize": 13,
                },
                Object {
                  "textAlign": "right",
                },
              ]
            }
          >
            submitKeyText
          </Text>
        </View>
      </View>
      <View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#FFFFFF",
              },
              Object {
                "backgroundColor": "#CCCCCC",
              },
              Object {
                "width": 750,
              },
            ]
          }
        >
          <RCTPicker
            items={
              Array [
                Object {
                  "label": "0",
                  "textColor": undefined,
                  "value": 0,
                },
              ]
            }
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            selectedIndex={0}
            style={
              Array [
                Object {
                  "height": 216,
                },
                undefined,
              ]
            }
          />
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
